File: android/word/src/main/java/com/dreampany/word/data/source/repository/WordRepository.java
/*
package com.dreampany.word.data.source.repository;

import android.graphics.Bitmap;
import com.dreampany.frame.data.model.State;
import com.dreampany.frame.data.source.repository.Repository;
import com.dreampany.frame.misc.*;
import com.dreampany.frame.util.DataUtil;
import com.dreampany.word.data.misc.WordMapper;
import com.dreampany.word.data.model.Word;
import com.dreampany.word.data.source.api.WordDataSource;
import com.dreampany.word.data.source.pref.Pref;
import io.reactivex.Flowable;
import io.reactivex.Maybe;

import javax.inject.Inject;
import javax.inject.Singleton;

import java.util.ArrayList;
import java.util.List;

*/
/**
 * Created by Hawladar Roman on 2/9/18.
 * Dreampany Ltd
 * dreampanymail@gmail.com
 *//*

@Singleton
public class WordRepository extends Repository<String, Word> implements WordDataSource {

    private final WordMapper mapper;
    private final Pref pref;
    private final WordDataSource assets;
    private final WordDataSource room;
    private final WordDataSource firestore;
    private final WordDataSource remote;
    private final WordDataSource vision;

    @Inject
    WordRepository(RxMapper rx,
                   ResponseMapper rm,
                   WordMapper mapper,
                   Pref pref,
                   @Assets WordDataSource assets,
                   @Room WordDataSource room,
                   @Firestore WordDataSource firestore,
                   @Remote WordDataSource remote,
                   @Vision WordDataSource vision) {
        super(rx, rm);
        this.mapper = mapper;
        this.pref = pref;
        this.assets = assets;
        this.room = room;
        this.firestore = firestore;
        this.remote = remote;
        this.vision = vision;
    }

    @Override
    public Word getTodayItem() {
        return null;
    }

    @Override
    public Maybe<Word> getTodayItemRx() {
*/
/*        Maybe<Word> room = getTodayFromRoom();
        Maybe<Word> generated = saveState(this.generateToday(), ItemState.TODAY);
        Maybe<Word> single = concatSingleFirstRx(room, generated);
        return single
                .filter(item -> item != null)
                .map(word -> {
*//*
*/
/*                    if (!hasState(word, ItemState.TODAY, ItemSubstate.FULL)) {
                        return getItemRx(word.getWord()).blockingGet();
                    }*//*
*/
/*
                    return word;
                });*//*

        return Maybe.empty();
    }

    @Override
    public boolean isEmpty() {
        return room.isEmpty();
    }

    @Override
    public Maybe<Boolean> isEmptyRx() {
        return Maybe.fromCallable(this::isEmpty);
    }

    @Override
    public int getCount() {
        return room.getCount();
    }

    @Override
    public Maybe<Integer> getCountRx() {
        return room.getCountRx();
    }

    @Override
    public boolean isExists(Word word) {
        boolean exists = room.isExists(word);
        if (!exists) {
            exists = assets.isExists(word);
        }
        return exists;
    }

    @Override
    public Maybe<Boolean> isExistsRx(Word word) {
        return Maybe.fromCallable(() -> isExists(word));
    }

    @Override
    public long putItem(Word word) {
        return room.putItem(word);
    }

    @Override
    public Maybe<Long> putItemRx(Word word) {
        return null;
    }

    @Override
    public List<Long> putItems(List<? extends Word> words) {
        return room.putItems(words);
    }

    @Override
    public Maybe<List<Long>> putItemsRx(List<? extends Word> words) {
        return room.putItemsRx(words);
    }

    @Override
    public int delete(Word word) {
        return 0;
    }

    @Override
    public Maybe<Integer> deleteRx(Word word) {
        return null;
    }

    @Override
    public List<Long> delete(List<? extends Word> words) {
        return null;
    }

    @Override
    public Maybe<List<Long>> deleteRx(List<? extends Word> words) {
        return null;
    }

    @Override
    public Word getItem(String id) {
        return room.getItem(id);
    }

    @Override
    public Maybe<Word> getItemRx(String id) {
        return null;
    }

    @Override
    public ArrayList<Word> getItems() {
        return null;
    }

    @Override
    public Maybe<List<Word>> getItemsRx() {
        Maybe<List<Word>> assets = getAssetsItemsIfRx();
        Maybe<List<Word>> room = this.room.getItemsRx();
        return concatLastRx(assets, room);
    }

    @Override
    public List<Word> getItems(int limit) {
        return null;
    }

    @Override
    public Maybe<List<Word>> getItemsRx(int limit) {
        return null;
    }

    @Override
    public Word getItem(String word, boolean full) {
        Word result = room.getItem(word, full);
        if (result == null) {
            result = assets.getItem(word, full);
        }
        return result;
    }

    @Override
    public Maybe<Word> getItemRx(String word, boolean full) {
*/
/*        Maybe<Word> room = fullRoom(word);
        Maybe<Word> firestore = saveRoom(this.firestore.getItemRx(word));
        Maybe<Word> remote = saveRoomFirestore(this.remote.getItemRx(word));
        return concatSingleFirstRx(room, firestore, remote);*//*

        return Maybe.empty();
    }

    @Override
    public List<Word> getItems(List<String> ids) {
        return room.getItems(ids);
    }

    @Override
    public Maybe<List<Word>> getItemsRx(List<String> ids) {
        return room.getItemsRx(ids);
    }

    @Override
    public List<Word> getSearchItems(String query, int limit) {
        return room.getSearchItems(query, limit);
    }

    @Override
    public List<Word> getCommonItems() {
        return assets.getCommonItems();
    }

    @Override
    public List<Word> getAlphaItems() {
        return assets.getAlphaItems();
    }

    @Override
    public Maybe<List<Word>> getItemsRx(Bitmap bitmap) {
        return null;
    }

    @Override
    public List<String> getRawWords() {
        return room.getRawWords();
    }

    @Override
    public Maybe<List<String>> getRawWordsRx() {
        Maybe<List<String>> assets = this.assets.getRawWordsRx();
        Maybe<List<String>> room = this.room.getRawWordsRx();
        return concatFirstOfStringRx(room, assets);
    }


    public Word getRoomItem(String word, boolean full) {
        return room.getItem(word, full);
    }

    public Word getFirestoreItem(String word, boolean full) {
        return firestore.getItem(word, full);
    }

    public Word getRemoteItem(String word, boolean full) {
        return remote.getItem(word, full);
    }

    public long putRoomItem(Word word) {
       return room.putItem(word);
    }

    public long putFirestoreItem(Word word) {
        return firestore.putItem(word);
    }

    public long putRemoteItem(Word word) {
        return remote.putItem(word);
    }

    public Word getItemOfMapper(String word) {
        return mapper.toItem(word);
    }

    public Word getItemOf(String word, boolean full) {
        Word result = room.getItem(word, full);
        if (result == null) {
            result = getItemOfMapper(word);
        }
        return result;
    }

    */
/* private api *//*

    private Maybe<List<Word>> getAssetsItemsIfRx() {
        return Maybe.fromCallable(() -> {
            if (getCount() == 0) {
                Maybe<List<Word>> maybe = saveRoomOfItems(assets.getItemsRx());
                return maybe.blockingGet();
            }
            return null;
        });
    }

    private Maybe<List<Word>> getItemsByStatesRx(List<State> items) {
        return Flowable.fromIterable(items)
                .map(item -> mapper.toItem(item, room))
                .toList()
                .toMaybe();
    }


    private Maybe<List<Word>> saveRoomOfItems(Maybe<List<Word>> source) {
        return source
                .filter(items -> !(DataUtil.isEmpty(items)))
                .doOnSuccess(words -> {
                    rx.compute(putItemsRx((ArrayList<Word>) words)).subscribe();
                });
    }

*/
/*    private Maybe<Word> getTodayFromRoom() {
        String type = ItemType.WORD.name();
        String subtype = ItemSubtype.DEFAULT.name();
        String state = ItemState.TODAY.name();
        long from = TimeUtil.startOfDay();
        long to = TimeUtil.endOfDay() - 1;
        return stateRepo.getItemOrderByRx(type, subtype, state, from, to)
                .map(item -> {
                    if (item != null) {
                        return mapper.toItem(item, room);
                    }
                    return null;
                });
    }*//*


*/
/*    private Maybe<Word> generateToday() {
        String type = ItemType.WORD.name();
        String subtype = ItemSubtype.DEFAULT.name();
        String state = ItemState.TODAY.name();
        return stateRepo.getItemNotStateOrderByRx(type, subtype, state)
                .map(item -> {
                    if (item != null) {
                        return mapper.toItem(item, room);
                    }
                    return null;
                });
    }*//*

}
*/

File: android/word/src/main/kotlin/com/dreampany/word/data/source/repository/WordRepository.kt
package com.dreampany.word.data.source.repository

import android.graphics.Bitmap
import com.dreampany.framework.data.source.repository.RepositoryKt
import com.dreampany.framework.misc.*
import com.dreampany.framework.util.DataUtil
import com.dreampany.word.data.misc.WordMapper
import com.dreampany.word.data.model.Word
import com.dreampany.word.data.source.api.WordDataSource
import com.dreampany.word.data.source.pref.Pref
import io.reactivex.Maybe
import java.util.*
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Created by roman on 2019-07-13
 * Copyright (c) 2019 bjit. All rights reserved.
 * hawladar.roman@bjitgroup.com
 * Last modified $file.lastModified
 */
@Singleton
class WordRepository @Inject constructor(
    rx: RxMapper,
    rm: ResponseMapper,
    val pref: Pref,
    val mapper: WordMapper,
    @Assets val assets: WordDataSource,
    @Room val room: WordDataSource,
    @Firestore val firestore: WordDataSource,
    @Remote val remote: WordDataSource,
    @Vision val vision: WordDataSource
) : RepositoryKt<String, Word>(rx, rm), WordDataSource {
    override fun isExists(word: String): Boolean {
        return assets.isExists(word)
    }

    override fun getTodayItem(): Word? {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    override fun getTodayItemRx(): Maybe<Word> {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    override fun getItem(word: String, full: Boolean): Word? {
        var result = room.getItem(word, full)
        if (result == null) {
            result = assets.getItem(word, full)
        }
        if (result == null) {

        }
        return result
    }

    override fun deleteRx(t: Word): Maybe<Int> {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    override fun getItem(id: String): Word? {
        return room.getItem(id)
    }

    override fun getItemRx(word: String, full: Boolean): Maybe<Word> {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    override fun getItemRx(id: String): Maybe<Word> {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    override fun getItems(ids: List<String>): List<Word> {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    override fun getItems(): List<Word> {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    override fun getItems(limit: Int): List<Word> {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    override fun getItemsRx(ids: List<String>): Maybe<List<Word>> {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    override fun getItemsRx(bitmap: Bitmap): Maybe<List<Word>> {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    override fun getItemsRx(): Maybe<List<Word>> {
        val assets = getAssetsItemsIfRx()
        val room = this.room.getItemsRx()
        return concatLastRx(assets, room)
    }

    override fun getItemsRx(limit: Int): Maybe<List<Word>> {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    override fun getSearchItems(query: String, limit: Int): List<Word> {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    override fun getCommonItems(): List<Word>? {
        return assets.getCommonItems();
    }

    override fun getAlphaItems(): List<Word>? {
        return assets.getAlphaItems()
    }

    override fun getRawWords(): List<String>? {
        return room.getRawWords()
    }

    override fun getRawWordsRx(): Maybe<List<String>> {
        val assets = this.assets.getRawWordsRx()
        val room = this.room.getRawWordsRx()
        return concatFirstOfStringRx(room, assets)
    }

    override fun isEmpty(): Boolean {
        return room.isEmpty()
    }

    override fun isEmptyRx(): Maybe<Boolean> {
        return Maybe.fromCallable({ this.isEmpty() })
    }

    override fun getCount(): Int {
        return room.getCount()
    }

    override fun getCountRx(): Maybe<Int> {
        return room.getCountRx()
    }

    override fun isExists(t: Word): Boolean {
        var exists = room.isExists(t)
        if (!exists) {
            exists = assets.isExists(t)
        }
        return exists
    }

    override fun isExistsRx(t: Word): Maybe<Boolean> {
        return Maybe.fromCallable { isExists(t) }
    }

    override fun putItem(t: Word): Long {
        return room.putItem(t)
    }

    override fun putItemRx(t: Word): Maybe<Long> {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    override fun putItems(ts: List<Word>): List<Long>? {
        return room.putItems(ts)
    }

    override fun putItemsRx(ts: List<Word>): Maybe<List<Long>> {
        return room.putItemsRx(ts)
    }

    override fun delete(t: Word): Int {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    override fun delete(ts: List<Word>): List<Long> {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    override fun deleteRx(ts: List<Word>): Maybe<List<Long>> {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    fun getRoomItem(word: String, full: Boolean): Word? {
        return room.getItem(word, full)
    }

    fun getFirestoreItem(word: String, full: Boolean): Word? {
        return firestore.getItem(word, full)
    }

    fun getRemoteItem(word: String, full: Boolean): Word? {
        return remote.getItem(word, full)
    }

    fun putRoomItem(word: Word): Long {
        return room.putItem(word)
    }

    fun putFirestoreItem(word: Word): Long {
        return firestore.putItem(word)
    }

    fun putRemoteItem(word: Word): Long {
        return remote.putItem(word)
    }

    fun getItemOfMapper(word: String): Word? {
        return mapper.toItem(word)
    }

    fun getItemOf(word: String, full: Boolean): Word? {
        var result = room.getItem(word, full)
        if (result == null) {
            result = getItemOfMapper(word)
        }
        return result
    }

    /* private api */
    private fun getAssetsItemsIfRx(): Maybe<List<Word>> {
        return Maybe.fromCallable {
            if (getCount() == 0) {
                val maybe = saveRoomOfItems(assets.getItemsRx())
                maybe.blockingGet()
            }
            null
        }
    }

/*    private fun getItemsByStatesRx(items: List<State>): Maybe<List<Word>> {
        return Flowable.fromIterable(items)
            .map { item -> mapper.toItem(item, room) }
            .toList()
            .toMaybe()
    }*/


    private fun saveRoomOfItems(source: Maybe<List<Word>>): Maybe<List<Word>> {
        return source
            .filter { items -> !DataUtil.isEmpty(items) }
            .doOnSuccess { words -> rx.compute(putItemsRx(words as ArrayList<Word>)).subscribe() }
    }

    /*    private Maybe<Word> getTodayFromRoom() {
        String type = ItemType.WORD.name();
        String subtype = ItemSubtype.DEFAULT.name();
        String state = ItemState.TODAY.name();
        long from = TimeUtil.startOfDay();
        long to = TimeUtil.endOfDay() - 1;
        return stateRepo.getItemOrderByRx(type, subtype, state, from, to)
                .map(item -> {
                    if (item != null) {
                        return mapper.toItem(item, room);
                    }
                    return null;
                });
    }*/

/*    private Maybe<Word> generateToday() {
        String type = ItemType.WORD.name();
        String subtype = ItemSubtype.DEFAULT.name();
        String state = ItemState.TODAY.name();
        return stateRepo.getItemNotStateOrderByRx(type, subtype, state)
                .map(item -> {
                    if (item != null) {
                        return mapper.toItem(item, room);
                    }
                    return null;
                });
    }*/
}